### PR TITLE
Change Compilation.CommonSyntaxTrees to use ImmutableArray

### PR DESCRIPTION
**Pull request by HurricanKai**
fixes https://github.com/dotnet/roslyn/issues/50617

### CODE REVIEW

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 1]
```diff
@@ -438,7 +438,7 @@ internal static bool IsValidHostObjectType(Type type)
         /// Gets the syntax trees (parsed from source code) that this compilation was created with.
         /// </summary>
         public IEnumerable<SyntaxTree> SyntaxTrees { get { return CommonSyntaxTrees; } }
-        protected abstract IEnumerable<SyntaxTree> CommonSyntaxTrees { get; }
+        protected abstract ImmutableArray<SyntaxTree> CommonSyntaxTrees { get; }
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs:441]: So this doesn't make anything worse, but it also doesn't make things significantly _better_. Every call to `Compilation.SyntaxTrees` is still going to be boxing an `ImmutableArray<SyntaxTree>` for the return value. Ideally we can find a solution that also removes the boxing penalty from `SyntaxTrees`.

Comment (333fred) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:441]: Attempting to change the public API is something that I'd prefer we didn't try to leave to a public contibutor :). This addresses the open issue, and we can look at making it better more deeply as a follow up.

**Review by 333fred — APPROVED** [review]

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review.

**Review by sharwell — APPROVED** [review]

File: src/Compilers/Core/Portable/Compilation/Compilation.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Compilation/Compilation.cs:441]: > So this doesn't make anything worse, but it also doesn't make things significantly better. Every call to Compilation.SyntaxTrees is still going to be boxing an ImmutableArray<SyntaxTree> for the return value. Ideally we can find a solution that also removes the boxing penalty from SyntaxTrees.

Perhaps we can add a different property that return an array?

---
In reply to: [595314226](https://github.com/dotnet/roslyn/pull/51905#discussion_r595314226) [](ancestors = 595314226)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 1)

**Comment by 333fred** [comment]
Thanks @HurricanKai!

**Pull request merged**